Disaster Recovery and Geo-Distribution Documentation

CarlosSardo opened the pull request:
More info here: https://github.com/Azure/azure-functions-durable-extension/issues/89

azurecla commented:
Hi __@SoubhagyaDash__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (sdash). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

alancameronwills commented:
#sign-off

PRmerger commented:
@SoubhagyaDash : Thanks for your contribution to the Azure documentation! The author, @soubhagyadash, has been notified to review your proposed change.

SoubhagyaDash commented:
#sign-off please push this along.

SoubhagyaDash commented:
#sign-off

PRmerger commented:
Only the author of the article can use the sign off comment to add the ready-to-merge label.

CarlosSardo committed:
Create Disaster Recovery and Geo-Distribution Documentation file

CarlosSardo committed:
Update TOC.yml with Disaster Recovery and Geo-Distribution Documentation

More details here: https://github.com/Azure/azure-functions-durable-extension/issues/89

MS-Santi committed:
First stab at the article

CarlosSardo committed:
Improved descriptions and markdown formatting

CarlosSardo committed:
extra comments

MS-Santi committed:
Added text for Scenario 1

MS-Santi committed:
Added text for the overview and Scenario 1

MS-Santi committed:
Overview and Scenario 1

MS-Santi committed:
Formatted using .md markup

CarlosSardo committed:
Added scenarios illustrations

MS-Santi committed:
Started with Scenario 2

MS-Santi committed:
Merge branch 'af-durable-geo-story' of https://github.com/CarlosSardo/azure-docs into af-durable-geo-story

CarlosSardo committed:
removed hardcoded URIs

MS-Santi committed:
Scenarios 2 and 3

MS-Santi committed:
Merge branch 'af-durable-geo-story' of https://github.com/CarlosSardo/azure-docs into af-durable-geo-story

MS-Santi committed:
Finished initial version of the document

CarlosSardo committed:
added missing image and notes

CarlosSardo committed:
resized images 780px

CarlosSardo committed:
resized images 500 px width

MS-Santi committed:
Corrected punctuation; Removed old comments

MS-Santi committed:
Merge branch 'af-durable-geo-story' of https://github.com/CarlosSardo/azure-docs into af-durable-geo-story

MS-Santi committed:
Try to center diagram for scenario 2

MS-Santi committed:
try 2

MS-Santi committed:
Try 3

CarlosSardo committed:
removed html from markdown

CarlosSardo committed:
added reference to Designing Highly Available Applications using RA-GRS

cgillum commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+## Overview
+In Azure Durable Functions, all state is persisted in Azure Storage. A [task hub](durable-functions-task-hubs.md) is a logical container for Azure Storage resources that are used for orchestrations. Orchestrator and activity functions can only interact with each other when they belong to the same task hub.
+The scenarios described in this article propose deployment options to increase availability and minimize downtime during disaster recovery activities.
+It is important to notice that these scenarios are based on an Active-Passive configurations, since they are guided by the usage of Azure Storage.
```
Is there an article that describes "Active-Passive configurations" that you can link to?

tdykstra replied:
Link to something about Active-Passive configurations as cgillum suggests, but also explain briefly so people don't have to click the link to understand the point being made.

CarlosSardo replied:
check

cgillum commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+>
+
+## Scenario 1 - Load Balanced Compute with Shared Storage
+In the event of problems of the compute service in Azure, the Function App could be affected. To minimize the possibility of such downtime, this scenario uses two instances of the Function App deployed to different regions. The underlying storage account and task hub are created in the main region, and it is shared by both instances.
```
Some feedback on this first paragraph:
* I would avoid capitalizing "Function App". Just "function app" in lowercase is good.
* Maybe "compute infrastructure" instead of "compute service"? "Service" sounds like it's a product named "Compute".
* "could be affected" is a little vague. Can you be more specific? For example, "the function app may become unavailable" or "the function app may be unable to run".

tdykstra replied:
problems of --> problems with

CarlosSardo replied:
check

cgillum commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+
+## Scenario 1 - Load Balanced Compute with Shared Storage
+In the event of problems of the compute service in Azure, the Function App could be affected. To minimize the possibility of such downtime, this scenario uses two instances of the Function App deployed to different regions. The underlying storage account and task hub are created in the main region, and it is shared by both instances.
+Traffic Manager is configured to detect problems in the main site and automatically redirect traffic to the healthy Function App in the failover region.
```
* Traffic Manager --> [Azure Traffic Manager](https://azure.microsoft.com/en-us/services/traffic-manager/) (with embedded link)
* site --> function app
* healthy Function App --> secondary function app (?)

CarlosSardo replied:
check

cgillum commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+## Scenario 1 - Load Balanced Compute with Shared Storage
+In the event of problems of the compute service in Azure, the Function App could be affected. To minimize the possibility of such downtime, this scenario uses two instances of the Function App deployed to different regions. The underlying storage account and task hub are created in the main region, and it is shared by both instances.
+Traffic Manager is configured to detect problems in the main site and automatically redirect traffic to the healthy Function App in the failover region.
+Because Function App the failover region shares the same Azure Storage account and Task Hub, the state of the functions is not lost, and word can resume normally. Once the health is restored in the main region, Traffic Manager will start routing requests to that instance of the Function App automatically.
```
* There are a few grammar mistakes in this first sentence.
* We actually advise against sharing the same storage account and task hub across multiple function apps. In this case it will work okay as long as the two function apps are exact copies of each other. However, background processing will be distributed across both of them since they will compete for messages on the same queues. This could be problematic due to the egress costs. A possible mitigation for this would be to disable the secondary function app when it is not needed, but this will require manual failover steps to re-enable it.

CarlosSardo replied:
check

cgillum commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+- This scenario covers outages at the compute layer, but the storage account continues to be the single point of failure for the Function App. If there is a Storage outage, the application suffers a downtime.
+- If the FunctionApp is failed over, there will be increased latency since it will access its storage account across regions.
+- Accessing the storage service from a different region where it is located incurs in higher cost due to network egress traffic.
+- This scenario depends on Traffic Manager. Considering [how Traffic Manager works](../traffic-manager/traffic-manager-overview#how-traffic-manager-works), it may be some time until a client application that consumes a Durable Function needs to re-query the FunctionApp address from Traffic Manager. 
```
FunctionApp --> function app

CarlosSardo replied:
check

cgillum commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+
+
+## Scenario 2 - Load Balanced Compute with Regional Storage
+The previous proposed scenario covers only the case of failure in the compute services. If the storage service fails, it will result in an outage of the FunctionApp.
```
The previous proposed scenario --> The previous**ly** proposed scenario

tdykstra replied:
Or "The preceding scenario..."

CarlosSardo replied:
check

cgillum reviewed (changes requested):
Here is some partial feedback. I'll continue to add additional feedback. We should also look into adding this to **azure-docs-pr** so that we can run automation against your document that will catch spelling mistakes, and automatically evaluate grammar, clarity, etc.

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+
+
+>[!NOTE]
+>All function apps that are redundantly deployed, must share the same function access keys in the case of being activated via HTTP. The Functions Runtime exposes a [management API](https://github.com/Azure/azure-functions-host/wiki/Key-management-API) that enables consumers to programmatically add, delete, and update function keys.
```
Why is this note here?  Its relation to the overview, and the reason for formatting it as a note, is not clear.

Link to the docs.ms.com article for the management API rather than the wiki.

CarlosSardo replied:
We kept the note, so that we put emphasis on the importance of the scenario basics.

@tdykstra where can we find it in docs?

tdykstra replied:
Sorry, my mistake, I thought I remembered seeing such a doc but can't find it now.

@ggailey777 Do you know if we cover the material in https://github.com/Azure/azure-functions-host/wiki/Key-management-API anywhere in our docs?

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+![Diagram showing scenario 1.](media/durable-functions-disaster-recovery-geo-distribution/durable-functions-geo-scenario01.png)
+
+There are several benefits when using this deployment scenario:
+- In the case of a failure in the compute layer, work can resume in the failover region without state loss.
```
what does "compute layer" refer to?  If function apps are a compute service, maybe the failover function app would also not be working.  This scenario addresses a regional failure of compute infrastructure, not a compute layer failure it seems to me.

CarlosSardo replied:
check (modified to compute infrastructure)

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+- Traffic Manager automatically re-establishes traffic to the main Function App after the outage has been corrected.
+
+However, consider the following when using this scenario:
+- If the Function App is deployed using a dedicated AppService plan, there is an increased cost by replicating the compute instance in the failover datacenter.
```
there is an increased cost by replicating the compute instance in the failover datacenter.

-->

replicating the compute instance in the failover datacenter increases costs.

CarlosSardo replied:
check

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+- If the Function App is deployed using a dedicated AppService plan, there is an increased cost by replicating the compute instance in the failover datacenter.
+- This scenario covers outages at the compute layer, but the storage account continues to be the single point of failure for the Function App. If there is a Storage outage, the application suffers a downtime.
+- If the FunctionApp is failed over, there will be increased latency since it will access its storage account across regions.
+- Accessing the storage service from a different region where it is located incurs in higher cost due to network egress traffic.
```
incurs in higher cost --> incurs higher cost

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+This approach adds improvements on the previous scenario:
+- In the case of failure, failover to the failover regions is still managed by Traffic Manager. However, because the FunctionApp relies on its own storage account, the durable functions continue to work.
+- During a failover, there is no additional latency in the failover region, since the FunctionApp and the storage account are co-located.
+- Failure of the storage layer will cause failures in the durable functions, which, in turn, will trigger a redirection to the failover region. Again, since the FunctionApp and storage are isolated per region, the durable functions will continue to work.
```
> Failure of the storage layer

This only addresses a regional failure, not the storage layer in general.

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+- Failure of the storage layer will cause failures in the durable functions, which, in turn, will trigger a redirection to the failover region. Again, since the FunctionApp and storage are isolated per region, the durable functions will continue to work.
+ 
+Important considerations for this scenario:
+- If the Function App is deployed using a dedicated AppService plan, there is an increased cost by replicating the compute instance in the failover datacenter.
```
AppService plan --> App Service plan

cost by replicating --> see earlier suggestion

CarlosSardo replied:
check

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+## Scenario 3 - Load Balanced Compute with GRS Shared Storage
+This scenario is a modification over the first scenario, implementing a shared storage account. The main difference that the storage account is created with geo-replication enabled.
+Functionally, this scenario provides the same advantages as Scenario 1, but it enables additional data recovery advantages:
+- Geo-redundant storage (GRS) and Read-access GRS (RA-GRS) maximize availability for your storage account. Additionally, RA-GRS provides read-only access to the data in the secondary location, in addition to geo-replication across two regions.
```
The sentence beginning with Additionally is confusing.  Geo-location across two regions implies you have access to the storage in the two regions, so why is read-only access "additional"?

CarlosSardo replied:
check (removed the comment)

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+Functionally, this scenario provides the same advantages as Scenario 1, but it enables additional data recovery advantages:
+- Geo-redundant storage (GRS) and Read-access GRS (RA-GRS) maximize availability for your storage account. Additionally, RA-GRS provides read-only access to the data in the secondary location, in addition to geo-replication across two regions.
+- If there is a region outage of the storage service, one of the possibilities is that the datacenter operations determine that storage must be failed over to the secondary region. In this case, storage account access will be redirected transparently to the geo-replicated copy of the storage account, without user intervention.
+- In this case, state of the durable functions will be preserved up to the last replication of the storage account.
```
> up to the last replication of the storage account.

Which happens how often?

CarlosSardo replied:
there is no real SLA on this "storage rep lag" story. therefore, we added: "which occurs every few minutes"

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+As with the other scenarios, there are important considerations:
+- Failover to the replica is done by datacenter operators and it may take some time. Until that time, the FunctionApp will suffer an outage.
+- There is an increased cost for using geo-replicated storage accounts.
+- GRS occurs asynchronously. It is possible that some or the latest transactions are lost because of the latency of the replication process.
```
It is possible that some or the latest transactions are lost --> Some of the latest transactions might be lost

tdykstra commented on articles/azure-functions/durable-functions-disaster-recovery-geo-distribution.md:
```diff
@@ -0,0 +1,81 @@
+
+You can read more about [Designing Highly Available Applications using RA-GRS](https://docs.microsoft.com/en-us/azure/storage/common/storage-designing-ha-apps-with-ragrs)
+
+![Diagram showing scenario 3.](media/durable-functions-disaster-recovery-geo-distribution/durable-functions-geo-scenario03.png)
```
At end need `## Next steps` section with "For more information, see ..." links

for example, the scenarios talk about configuring traffic manager, so it would be good to link to the documentation that explains how to do that.

CarlosSardo replied:
check

tdykstra reviewed (commented):
A general question is whether this belongs in Durable Functions documentation or general Functions documentation.  The scenarios appear to apply to function apps in general, not just function apps running durable functions.

MS-Santi committed:
Update per ktoliver request for merging to occur
